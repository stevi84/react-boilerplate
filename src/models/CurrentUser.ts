import { Components } from '../apis/api';

export type CurrentUser = Components.Schemas.CurrentUser;
export type Role = Components.Schemas.Role;
export const getEmptyCurrentUser = (): CurrentUser => ({
  loginName: '',
  firstName: '',
  surName: '',
  mail: '',
  roles: [],
});
