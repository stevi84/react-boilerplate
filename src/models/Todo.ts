import { Components } from '../apis/api';

export type Todo = Components.Schemas.Todo;

export const getEmptyTodo = (): Todo => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  owner: '',
  dueDate: '',
  description: '',
  completed: false,
});
