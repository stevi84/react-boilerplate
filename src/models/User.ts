import { Components } from '../apis/api';

export type User = Components.Schemas.User;

export const getEmptyUser = (): User => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  name: '',
  dateOfBirth: '',
  size: 0,
  weight: 0,
  email: '',
  phone: '',
});
