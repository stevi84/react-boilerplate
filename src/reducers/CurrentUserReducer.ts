import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from './Store';
import { CurrentUser, getEmptyCurrentUser } from '../models/CurrentUser';

export const currentUserInitialState: CurrentUser = getEmptyCurrentUser();

const currentUserSlice = createSlice({
  name: 'currentUser',
  initialState: currentUserInitialState,
  reducers: {
    setCurrentUser(state, action: PayloadAction<CurrentUser>) {
      return action.payload;
    },
  },
});

const { actions, reducer } = currentUserSlice;
export const { setCurrentUser } = actions;
export { reducer as currentUserReducer };

export const currentUserSelector = (state: RootState): CurrentUser => state?.currentUser || currentUserInitialState;
