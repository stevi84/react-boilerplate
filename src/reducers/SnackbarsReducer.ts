import { PrepareAction, createAction, createReducer } from '@reduxjs/toolkit';
import { OptionsObject, SnackbarKey, SnackbarMessage } from 'notistack';
import { RootState } from './Store';

// https://stackoverflow.com/questions/62754584/how-can-i-use-notistick-any-snackbar-with-redux-toolkit-and-react
// https://codesandbox.io/s/github/iamhosseindhv/notistack/tree/master/examples/redux-example?file=/redux/actions.js:0-150

interface Snackbar {
  message: SnackbarMessage;
  options?: OptionsObject;
  key: SnackbarKey;
  dismissed: boolean;
}

interface SnackbarState {
  notifications: Snackbar[];
}

export const enqueueSnackbar = createAction<PrepareAction<{ notification: Snackbar }>>(
  'ENQUEUE_SNACKBAR',
  (message: SnackbarMessage, options?: OptionsObject) => ({
    payload: {
      notification: {
        message,
        options,
        key: (options && options.key) || new Date().getTime() + Math.random(),
        dismissed: false,
      },
    },
  })
);

export const closeSnackbar = createAction<PrepareAction<{ key?: SnackbarKey; dismissAll: boolean }>>(
  'CLOSE_SNACKBAR',
  (key?: SnackbarKey) => ({
    payload: {
      dismissAll: !key, // dismiss all if no key has been defined
      key,
    },
  })
);

export const removeSnackbar = createAction<PrepareAction<{ key: SnackbarKey }>>(
  'REMOVE_SNACKBAR',
  (key: SnackbarKey) => ({
    payload: {
      key,
    },
  })
);

export const snackbarsInitialState: SnackbarState = {
  notifications: [],
};

export const snackbarsReducer = createReducer(snackbarsInitialState, (builder) =>
  builder
    .addCase(enqueueSnackbar, (state, action) => ({
      notifications: [...state.notifications, { ...action.payload.notification }],
    }))
    .addCase(closeSnackbar, (state, action) => ({
      notifications: state.notifications.map((notification) =>
        action.payload.dismissAll || notification.key === action.payload.key
          ? { ...notification, dismissed: true }
          : { ...notification }
      ),
    }))
    .addCase(removeSnackbar, (state, action) => ({
      notifications: state.notifications.filter((n) => n.key !== action.payload.key),
    }))
);

export const notificationsSelector = (state: RootState): Snackbar[] => state?.snackbar?.notifications || [];
