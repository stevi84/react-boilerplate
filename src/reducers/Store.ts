import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { todosReducer } from './TodosReducer';
import { apiCallsReducer } from './ApiCallsReducer';
import { snackbarsReducer } from './SnackbarsReducer';
import { usersReducer } from './UsersReducer';
import { currentUserReducer } from './CurrentUserReducer';

export const store = configureStore({
  reducer: {
    snackbar: snackbarsReducer,
    todos: todosReducer,
    users: usersReducer,
    apiCalls: apiCallsReducer,
    currentUser: currentUserReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
