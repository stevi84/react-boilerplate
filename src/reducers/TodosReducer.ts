import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Todo } from '../models/Todo';
import { RootState } from './Store';

export const todosInitialState: Todo[] = [];

const todosSlice = createSlice({
  name: 'todos',
  initialState: todosInitialState,
  reducers: {
    setTodos(state, action: PayloadAction<Todo[]>) {
      return action.payload;
    },
    addTodo(state, action: PayloadAction<Todo>) {
      return state.concat(action.payload);
    },
    changeTodo(state, action: PayloadAction<Todo>) {
      return state.map((todo) => (todo.id !== action.payload.id ? todo : action.payload));
    },
    removeTodo(state, action: PayloadAction<number>) {
      return state.filter((todo) => todo.id !== action.payload);
    },
  },
});

const { actions, reducer } = todosSlice;
export const { setTodos, addTodo, changeTodo, removeTodo } = actions;
export { reducer as todosReducer };

export const todosSelector = (state: RootState): Todo[] => state?.todos || [];
