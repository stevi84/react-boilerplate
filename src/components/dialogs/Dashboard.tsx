import { ButtonBase, Card, CardContent, Grid, Paper, Typography } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import People from '@mui/icons-material/People';
import EditNote from '@mui/icons-material/EditNote';
import { useEffect } from 'react';
import { useAppDispatch, useAppSelector } from '../../reducers/Store';
import { readTodos, todosSelector } from '../../reducers/TodosReducer';
import { isReadingSelector, isSubmittingSelector } from '../../reducers/ApiCallsReducer';
import { Todo } from '../../models/Todo';
import { Working } from '../common/Working';
import { MainLayout } from '../common/MainLayout';
import { User } from '../../models/User';
import { readUsers, usersSelector } from '../../reducers/UsersReducer';
import { useTranslation } from 'react-i18next';
import { Locale } from '../../globals/Translations';

export const Dashboard = () => {
  const todos: Todo[] = useAppSelector(todosSelector);
  const users: User[] = useAppSelector(usersSelector);
  const isReading: boolean = useAppSelector(isReadingSelector);
  const isSubmitting: boolean = useAppSelector(isSubmittingSelector);
  const { i18n } = useTranslation();
  const lang: Locale = i18n.language as Locale;
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (todos.length === 0) {
      dispatch(readTodos(lang));
    }
    if (users.length === 0) {
      dispatch(readUsers(lang));
    }
  }, []);

  return (
    <MainLayout>
      <Paper sx={{ margin: 1 }}>
        {!isReading && !isSubmitting && (
          <Grid container columnSpacing={2} sx={{ padding: 1 }}>
            <Grid item>
              <Card>
                <ButtonBase onClick={(e) => navigate('/todos')}>
                  <CardContent>
                    <EditNote fontSize="large" />
                    <Typography variant="h5" component="div">
                      Todos
                    </Typography>
                  </CardContent>
                </ButtonBase>
              </Card>
            </Grid>
            <Grid item>
              <Card>
                <ButtonBase onClick={(e) => navigate('/users')}>
                  <CardContent>
                    <People fontSize="large" />
                    <Typography variant="h5" component="div">
                      Users
                    </Typography>
                  </CardContent>
                </ButtonBase>
              </Card>
            </Grid>
          </Grid>
        )}
        {(isReading || isSubmitting) && <Working isReading={isReading} isSubmitting={isSubmitting} />}
      </Paper>
    </MainLayout>
  );
};
