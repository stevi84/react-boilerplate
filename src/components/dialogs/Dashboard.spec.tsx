import { describe, it, expect, vi } from 'vitest';
import { fireEvent, screen } from '@testing-library/react';
import { Dashboard } from './Dashboard';
import { todo1 } from '../../test/data/Todo';
import { user1 } from '../../test/data/User';
import { currentUserAdmin, currentUserEditor } from '../../test/data/CurrentUser';
import { renderWithProviders } from '../../test/Utils';
import { RootState } from '../../reducers/Store';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

const { navigateMock } = vi.hoisted(() => ({ navigateMock: vi.fn() }));
vi.mock('react-router-dom', () => ({ useNavigate: () => navigateMock }));

vi.mock('../../hooks/UseNotifier', () => ({ useNotifier: vi.fn() }));

const { dispatchMock } = vi.hoisted(() => ({ dispatchMock: vi.fn() }));
vi.mock('../../reducers/Store', async () => {
  const mod = await vi.importActual<typeof import('../../reducers/Store')>('../../reducers/Store');
  return {
    ...mod,
    useAppDispatch: () => dispatchMock.mockImplementation(mod.useAppDispatch()),
  };
});

vi.mock('../../thunks/TodosThunks', () => ({
  readTodos: () => ({ type: 'readTodos' }),
}));

vi.mock('../../thunks/UsersThunks', () => ({
  readUsers: () => ({ type: 'readUsers' }),
}));

vi.mock('../../thunks/CurrentUserThunks', () => ({
  readCurrentUser: () => ({ type: 'readCurrentUser' }),
}));

const initialState: Partial<RootState> = {
  todos: [todo1],
  users: [user1],
  apiCalls: { runningReads: 0, runningSubmits: 0 },
  currentUser: currentUserAdmin,
};

describe('Dashboard', () => {
  it('should equal saved snapshot', () => {
    const tree = renderWithProviders(<Dashboard />, { preloadedState: initialState }).asFragment();
    expect(tree).toMatchSnapshot();
  });

  it('should load todos if not present', () => {
    expect(dispatchMock.mock.calls.length).toEqual(0);
    renderWithProviders(<Dashboard />, { preloadedState: { ...initialState, todos: [] } });
    expect(dispatchMock.mock.calls.length).toEqual(1);
    expect(dispatchMock.mock.calls[0][0]).toEqual({ type: 'readTodos' });
  });

  it('should load users if not present', () => {
    expect(dispatchMock.mock.calls.length).toEqual(0);
    renderWithProviders(<Dashboard />, { preloadedState: { ...initialState, users: [] } });
    expect(dispatchMock.mock.calls.length).toEqual(1);
    expect(dispatchMock.mock.calls[0][0]).toEqual({ type: 'readUsers' });
  });

  it('should navigate to todos on button click', () => {
    renderWithProviders(<Dashboard />, { preloadedState: initialState });
    expect(navigateMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('EditNoteIcon'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/todos');
  });

  it('should navigate to users on button click', () => {
    renderWithProviders(<Dashboard />, { preloadedState: initialState });
    expect(navigateMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('PeopleIcon'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/users');
  });

  it('should navigate to settings on button click', () => {
    renderWithProviders(<Dashboard />, { preloadedState: initialState });
    expect(navigateMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('SettingsIcon'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/admin');
  });

  it('should show dashboard if not reading or submitting', () => {
    renderWithProviders(<Dashboard />, { preloadedState: initialState });
    expect(screen.getByText('todo')).toBeTruthy();
    expect(screen.queryByRole('progressbar')).toBeFalsy();
  });

  it('should show working if reading', () => {
    renderWithProviders(<Dashboard />, {
      preloadedState: { ...initialState, apiCalls: { runningReads: 1, runningSubmits: 0 } },
    });
    expect(screen.queryByText('todo')).toBeFalsy();
    expect(screen.getByRole('progressbar')).toBeTruthy();
  });

  it('should show working if submitting', () => {
    renderWithProviders(<Dashboard />, {
      preloadedState: { ...initialState, apiCalls: { runningReads: 0, runningSubmits: 1 } },
    });
    expect(screen.queryByText('todo')).toBeFalsy();
    expect(screen.getByRole('progressbar')).toBeTruthy();
  });

  it('should hide settings if not authorized', () => {
    renderWithProviders(<Dashboard />, {
      preloadedState: { ...initialState, currentUser: currentUserEditor },
    });
    expect(screen.queryByText('settings')).toBeFalsy();
  });
});
