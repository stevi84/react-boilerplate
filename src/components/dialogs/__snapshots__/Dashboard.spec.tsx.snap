// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dashboard > should equal saved snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-5czdux-MuiPaper-root-MuiAppBar-root"
    style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-95f9g5-MuiToolbar-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-1by7r3h-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
          />
        </svg>
      </button>
      <div
        class="MuiTypography-root MuiTypography-h6 css-173147z-MuiTypography-root"
      >
        React Boilerplate
      </div>
      <div
        class="MuiTypography-root MuiTypography-body1 css-kcbyr-MuiTypography-root"
      >
        Environment: test
      </div>
      <div
        class="MuiBox-root css-ece9u5"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-1b5lkxe-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-18pjc51-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
          >
            language
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1cprn2g-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-controls=":r0:"
              aria-expanded="false"
              aria-haspopup="listbox"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-15k6ek6-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              role="combobox"
              tabindex="0"
            >
              de
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="de"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-1u6jos5-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-kg5swy"
              >
                <span>
                  language
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-95f9g5-MuiToolbar-root"
  />
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1r8heb5-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1qo8szv-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-6fm6rz-MuiPaper-root-MuiCard-root"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <button
            class="MuiButtonBase-root css-10d1a0h-MuiButtonBase-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-19vn4a9-MuiSvgIcon-root"
                data-testid="EditNoteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 10h11v2H3zm0-2h11V6H3zm0 8h7v-2H3zm15.01-3.13.71-.71c.39-.39 1.02-.39 1.41 0l.71.71c.39.39.39 1.02 0 1.41l-.71.71zm-.71.71-5.3 5.3V21h2.12l5.3-5.3z"
                />
              </svg>
              <div
                class="MuiTypography-root MuiTypography-h5 css-ku7d13-MuiTypography-root"
              >
                todo
              </div>
            </div>
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-6fm6rz-MuiPaper-root-MuiCard-root"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <button
            class="MuiButtonBase-root css-10d1a0h-MuiButtonBase-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-19vn4a9-MuiSvgIcon-root"
                data-testid="PeopleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5s-3 1.34-3 3 1.34 3 3 3m-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5 5 6.34 5 8s1.34 3 3 3m0 2c-2.33 0-7 1.17-7 3.5V19h14v-2.5c0-2.33-4.67-3.5-7-3.5m8 0c-.29 0-.62.02-.97.05 1.16.84 1.97 1.97 1.97 3.45V19h6v-2.5c0-2.33-4.67-3.5-7-3.5"
                />
              </svg>
              <div
                class="MuiTypography-root MuiTypography-h5 css-ku7d13-MuiTypography-root"
              >
                user
              </div>
            </div>
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-6fm6rz-MuiPaper-root-MuiCard-root"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <button
            class="MuiButtonBase-root css-10d1a0h-MuiButtonBase-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-19vn4a9-MuiSvgIcon-root"
                data-testid="SettingsIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6"
                />
              </svg>
              <div
                class="MuiTypography-root MuiTypography-h5 css-ku7d13-MuiTypography-root"
              >
                settings
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
